### PR TITLE
Add plist parser for iOS CarPlayApp.plist file

### LINKED ISSUE #4155: Add plist parser for iOS CarPlayApp.plist file
I propose to add a plist parser for iOS CarPlayApp.plist. I will make a pull request for this parser.

### PR DESCRIPTION
**Pull request by studiawan**
## One line description of pull request
Add plist parser for iOS CarPlayApp.plist file

## Description:
I add/edit several files:
- Add ios_carplay to __init__.py
- Add plaso/parsers/plist_plugins/ios_carplay.py
- Add test_data/com.apple.CarPlayApp.plist
- Add tests/parsers/plist_plugins/ios_carplay.py

**Related issue (if applicable):** fixes #4155 

## Notes:
All contributions to Plaso undergo [code
review](https://github.com/log2timeline/l2tdocs/blob/main/process/Code%20review%20process.asciidoc). This makes sure
that the code has appropriate test coverage and conforms to the Plaso [style
guide](https://plaso.readthedocs.io/en/latest/sources/developer/Style-guide.html).

One of the maintainers will examine your code, and may request changes. Check off the items below in
order, and then a maintainer will review your code.

## Checklist:
  - [x] Automated checks (Travis, Codecov, Codefactor )pass
  - [x] No new [new dependencies](https://plaso.readthedocs.io/en/latest/sources/developer/Adding-a-new-dependency.html) are required or l2tdevtools has been updated
  - [x] Reviewer assigned

### CODE REVIEW

**Comment by joachimmetz** [comment]
@studiawan please fix the linter warnings https://github.com/log2timeline/plaso/runs/7375412875?check_suite_focus=true

**Comment by studiawan** [comment]
I am working on this issue. I am still reading the pylint manual. Thanks.

**Comment by joachimmetz** [comment]
This might be an useful reference then https://plaso.readthedocs.io/en/latest/sources/developer/Style-guide.html you can run pylint locally with conditions similar to the CI tests with `tox -elint`

**Comment by studiawan** [comment]
Hi @joachimmetz, I have fixed the linter warnings. Please have a look. Thanks.

File: plaso/parsers/plist_plugins/ios_carplay.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
+# -*- coding: utf-8 -*-
+"""Plist parser plugin for Car Play App plist files.
+   It contains history of opened apps on Car Play App.
```
Comment (joachimmetz) [inline plaso/parsers/plist_plugins/ios_carplay.py:3]: style nit: add empty line before second line of text and remove indentation

File: plaso/parsers/plist_plugins/ios_carplay.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+from plaso.containers import plist_event
+from plaso.containers import time_events
+from plaso.lib import definitions
+from plaso.parsers import plist
+from plaso.parsers.plist_plugins import interface
+
+
+class IOSCarPlayPlugin(interface.PlistPlugin):
+  """Plist parser plugin for IOS Car Play App plist files.
```
Comment (joachimmetz) [inline plaso/parsers/plist_plugins/ios_carplay.py:16]: style nit: put the closing quote on the same line

File: plaso/parsers/plist_plugins/ios_carplay.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+  """
+
+  NAME = 'ios_carplay'
+  DATA_TYPE = 'ios:carplay:history'
+  DATA_FORMAT = 'iOS Car Play App plist file'
+
+  PLIST_PATH_FILTERS = frozenset([
+    interface.PlistPathFilter('com.apple.CarPlayApp.plist')])
```
Comment (joachimmetz) [inline plaso/parsers/plist_plugins/ios_carplay.py:25]: style nit: use continuation indentation (4 spaces)

File: plaso/parsers/plist_plugins/ios_carplay.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+from plaso.parsers.plist_plugins import interface
+
+
+class IOSCarPlayPlugin(interface.PlistPlugin):
+  """Plist parser plugin for IOS Car Play App plist files.
+
+  """
+
+  NAME = 'ios_carplay'
+  DATA_TYPE = 'ios:carplay:history'
```
Comment (joachimmetz) [inline plaso/parsers/plist_plugins/ios_carplay.py:21]: Plugin does not need to define a data type, this should be done in an event data object

File: plaso/parsers/plist_plugins/ios_carplay.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+  DATA_FORMAT = 'iOS Car Play App plist file'
+
+  PLIST_PATH_FILTERS = frozenset([
+    interface.PlistPathFilter('com.apple.CarPlayApp.plist')])
+
+  PLIST_KEYS = frozenset(['CARRecentAppHistory'])
+
+  # pylint: disable=arguments-differ
+  def _ParsePlist(self, parser_mediator, match=None, **unused_kwargs):
+    """Extract relevant Car Play App entries.
```
Comment (joachimmetz) [inline plaso/parsers/plist_plugins/ios_carplay.py:31]: style nit: put the closing quote on the same line

File: plaso/parsers/plist_plugins/ios_carplay.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    """
+    recent_app_history = match.get('CARRecentAppHistory', {})
```
Comment (joachimmetz) [inline plaso/parsers/plist_plugins/ios_carplay.py:33]: style nit: don't use abbreviations like app, they don't translate well across an international audience

File: plaso/parsers/plist_plugins/ios_carplay.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    for parameter, value in recent_app_history.items():
+      event_data = plist_event.PlistTimeEventData()
```
Comment (joachimmetz) [inline plaso/parsers/plist_plugins/ios_carplay.py:35]: Define a data type specific EventData object

File: plaso/parsers/plist_plugins/ios_carplay.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+      event_data.root = '/CARRecentAppHistory'
```
Comment (joachimmetz) [inline plaso/parsers/plist_plugins/ios_carplay.py:36]: Why hard code this string?

File: plaso/parsers/plist_plugins/ios_carplay.py [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+      datetime_value = value
```
Comment (joachimmetz) [inline plaso/parsers/plist_plugins/ios_carplay.py:38]: why is this temporary variable needed?

File: plaso/parsers/plist_plugins/ios_carplay.py [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+      if datetime_value:
+        event_data.desc = parameter
```
Comment (joachimmetz) [inline plaso/parsers/plist_plugins/ios_carplay.py:40]: style nit: don't use abbreviations

File: plaso/parsers/plist_plugins/ios_carplay.py [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+        event_data.key = parameter
+
+        # Convert the floating point value to an integer.
+        # TODO: add support for the fractional part of
+        # the floating point value.
+        timestamp = int(datetime_value)
```
Comment (joachimmetz) [inline plaso/parsers/plist_plugins/ios_carplay.py:46]: what if this raises TypeError or ValueError ?

File: tests/parsers/plist_plugins/ios_carplay.py [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+        'recovery_warning')
+    self.assertEqual(number_of_warnings, 0)
+
+    # The order in which PlistParser generates events is nondeterministic
+    # hence we sort the events.
+    events = list(storage_writer.GetSortedEvents())
+
+    expected_event_values = {
+      'date_time': '2020-04-12 13:55:51',
+      'desc': 'com.apple.mobilecal'
```
Comment (joachimmetz) [inline tests/parsers/plist_plugins/ios_carplay.py:40]: style nit: put the closing bracket on the same line

File: plaso/parsers/plist_plugins/ios_carplay.py [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+      event_data = plist_event.PlistTimeEventData()
+      event_data.root = '/CARRecentAppHistory'
+
+      datetime_value = value
+      if datetime_value:
+        event_data.desc = parameter
+        event_data.key = parameter
+
+        # Convert the floating point value to an integer.
+        # TODO: add support for the fractional part of
```
Comment (joachimmetz) [inline plaso/parsers/plist_plugins/ios_carplay.py:44]: Please fix TODO

**Review by joachimmetz — CHANGES_REQUESTED** [review]
First round of comments, some style nits but a couple of structural issues

1. define a custom EventData object
2. catch exceptions generated by int(datetime_value)
3. fix the TODO, there is a PosixTimeInNanoseconds if you need high precision

File: plaso/parsers/plist_plugins/ios_carplay.py (hunk 7 above)
Comment (studiawan) [reply plaso/parsers/plist_plugins/ios_carplay.py:35]: Hi @joachimmetz what is `a data type specific EventData object`? Can you point an example from another plist parser?

File: plaso/parsers/plist_plugins/ios_carplay.py (hunk 8 above)
Comment (studiawan) [reply plaso/parsers/plist_plugins/ios_carplay.py:36]: To provide a root name for the dictionary that has datetime value. 
```
"CARRecentAppHistory": {
		"com.apple.mobilecal": 1586699751.255235,
		"com.apple.Music": 1586701543.8468399,
		"com.apple.Maps": 1586699753.1512132,
		"com.apple.mobilephone": 1586700310.690891,
		"com.apple.cardisplay.nowplaying": 1586701543.850651
	}
```

**Comment by joachimmetz** [comment]
FYI CI tests are failing with:

```
======================================================================
FAIL: testProcess (parsers.plist_plugins.ios_carplay.IOSCarPlayPluginTest)
Tests the Process function.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\projects\plaso\tests\parsers\plist_plugins\ios_carplay.py", line 42, in testProcess
    self.CheckEventValues(storage_writer, events[0], expected_event_values)
  File "C:\projects\plaso\.\tests\parsers\test_lib.py", line 203, in CheckEventValues
    self.assertEqual(value, expected_value, error_message)
AssertionError: None != 'com.apple.mobilecal' : event value: "desc" does not match expected value
----------------------------------------------------------------------
```

File: plaso/parsers/plist_plugins/ios_carplay.py (hunk 7 above)
Comment (joachimmetz) [reply plaso/parsers/plist_plugins/ios_carplay.py:32]: I see that plist parsers still use a generic PlistTimeEventData, I'll fix this

File: plaso/parsers/plist_plugins/ios_carplay.py (hunk 8 above)
Comment (joachimmetz) [reply plaso/parsers/plist_plugins/ios_carplay.py:33]: I see, this is a side effect of PlistTimeEventData

**Comment by joachimmetz** [comment]
@studiawan took another look. Looks like most of the comments I have are related to PlistTimeEventData, I'll consider this out of scope of this PR and have created https://github.com/log2timeline/plaso/issues/4231

**Review by joachimmetz — APPROVED** [review]
LGTM

**Pull request merged**